pkg/quota/image/imagestreamimport_evaluator.go

```diff
@@ -0,0 +1,220 @@
+// ImageStreamImport kind.
+func (c *imageStreamImportUsageComputer) Usage(object runtime.Object) kapi.ResourceList {
+	isi, ok := object.(*imageapi.ImageStreamImport)
+	if !ok {
+		return kapi.ResourceList{}
+	}
+
+	usage := map[kapi.ResourceName]resource.Quantity{
+		imageapi.ResourceImageStreamTags:   *resource.NewQuantity(0, resource.DecimalSI),
+		imageapi.ResourceImageStreamImages: *resource.NewQuantity(0, resource.DecimalSI),
+	}
+
+	if !isi.Spec.Import || (len(isi.Spec.Images) == 0 && isi.Spec.Repository == nil) {
+		return usage
+	}
+
+	// collect all the image references already tagged in a project
+	iss, err := c.osClient.ImageStreams(isi.Namespace).List(kapi.ListOptions{})
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to list image streams: %v", err))
+		return usage
+	}
+
+	for _, imageStream := range iss.Items {
+		c.ProcessImageStreamImages(&imageStream, false, func(ref string, inSpec, _ bool) error {
+			if !c.processedSpecRefs.Has(ref) && inSpec {
+				c.processedSpecRefs.Insert(ref)
+			}
+			if !c.processedRefs.Has(ref) {
+				c.processedRefs.Insert(ref)
+			}
+			return nil
+		})
+	}
+
+	// first process and remember individual images
+	imagesUsage := c.getImagesUsageIncrement(isi)
+	// second process a repository whose tags shall we import
+	repositoryUsage := c.getRepositoryUsageIncrement(isi)
+
+	return kquota.Add(imagesUsage, repositoryUsage)
+}
+
+func (c *imageStreamImportUsageComputer) getImagesUsageIncrement(isi *imageapi.ImageStreamImport) kapi.ResourceList {
+	specRefsIncrement := resource.NewQuantity(0, resource.DecimalSI)
+	allRefsIncrement := resource.NewQuantity(0, resource.DecimalSI)
+
+	for _, imageSpec := range isi.Spec.Images {
+		if imageSpec.From.Kind != "DockerImage" {
+			continue
+		}
+		ref, err := c.GetImageReferenceForObjectReference(isi.Namespace, &imageSpec.From)
+		if err != nil {
+
+		}
+		if !c.processedSpecRefs.Has(ref) {
+			c.processedSpecRefs.Insert(ref)
+			specRefsIncrement.Set(specRefsIncrement.Value() + 1)
+		}
+		if !c.processedRefs.Has(ref) {
+			c.processedRefs.Insert(ref)
+			allRefsIncrement.Set(allRefsIncrement.Value() + 1)
+		}
+	}
+
+	return map[kapi.ResourceName]resource.Quantity{
+		imageapi.ResourceImageStreamTags:   *specRefsIncrement,
+		imageapi.ResourceImageStreamImages: *allRefsIncrement,
+	}
+}
+
+// getRepositoryUsageIncrement returns a usage increment for ImageStreamImport.Spec.Repository. The repository
+// causes image importer to import all the repository tags in an image stream. We don't attempt to resolve
+// number of tags here. We just expect the worst case scenario where the max limit of import tags per
+// repository is reached and decrease it by the number of image references pointing to the repository that are
+// already tagged in a project.
+func (c *imageStreamImportUsageComputer) getRepositoryUsageIncrement(isi *imageapi.ImageStreamImport) kapi.ResourceList {
```

Comment:
My point is that these values *belong* to image streams, not image stream
import.  ISI is a virtual resource on top of the IS.  I would expect
replenishment and calculation of actual usage to happen on image streams,
while ISI is merely calculating the delta (that it would change) against
the usage calculated by IS.  I don't think we should start a policy of
tracking usage against virtual resources.

On Sat, Apr 9, 2016 at 2:41 PM, Michal Minar <notifications@github.com>
wrote:

> In pkg/quota/image/imagestreamimport_evaluator.go
> <https://github.com/openshift/origin/pull/8195#discussion_r59119389>:
>
> > +			allRefsIncrement.Set(allRefsIncrement.Value() + 1)
> > +		}
> > +	}
> > +
> > +	return map[kapi.ResourceName]resource.Quantity{
> > +		imageapi.ResourceImageStreamTags:   *specRefsIncrement,
> > +		imageapi.ResourceImageStreamImages: *allRefsIncrement,
> > +	}
> > +}
> > +
> > +// getRepositoryUsageIncrement returns a usage increment for ImageStreamImport.Spec.Repository. The repository
> > +// causes image importer to import all the repository tags in an image stream. We don't attempt to resolve
> > +// number of tags here. We just expect the worst case scenario where the max limit of import tags per
> > +// repository is reached and decrease it by the number of image references pointing to the repository that are
> > +// already tagged in a project.
> > +func (c *imageStreamImportUsageComputer) getRepositoryUsageIncrement(isi *imageapi.ImageStreamImport) kapi.ResourceList {
>
> No, the evaluation is done solely for the admission check.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8195/files/34b5aa7c22cb03e6c75e878c30ec0d7da02d5fd0#r59119389>
>